Zend/zend_frameless_function.h

```diff
@@ -0,0 +1,135 @@
+	if (Z_TYPE_P(arg ## arg_num) == IS_STRING) { \
+		dest_ht = NULL; \
+		dest_str = Z_STR_P(arg ## arg_num); \
+	} else if (EXPECTED(Z_TYPE_P(arg ## arg_num) == IS_ARRAY)) { \
+		dest_ht = Z_ARRVAL_P(arg ## arg_num); \
+		dest_str = NULL; \
+	} else { \
+		dest_ht = NULL; \
+		ZVAL_COPY(&str_tmp, arg ## arg_num); \
+		arg ## arg_num = &str_tmp; \
+		if (!zend_flf_parse_arg_str_slow(arg ## arg_num, &dest_str, arg_num)) { \
+			zend_wrong_parameter_type_error(arg_num, Z_EXPECTED_ARRAY_OR_STRING, arg ## arg_num); \
+			goto flf_clean; \
+		} \
+	}
+#define Z_FLF_PARAM_BOOL(arg_num, dest) \
+	if (!zend_parse_arg_bool_ex(arg ## arg_num, &dest, /* is_null */ NULL, /* null_check */ false, arg_num, /* frameless */ true)) { \
+		zend_wrong_parameter_type_error(arg_num, Z_EXPECTED_BOOL, arg ## arg_num); \
+		goto flf_clean; \
+	}
+#define Z_FLF_PARAM_LONG(arg_num, dest) \
+	if (!zend_parse_arg_long_ex(arg ## arg_num, &dest, /* is_null */ NULL, /* null_check */ false, arg_num, /* frameless */ true)) { \
+		zend_wrong_parameter_type_error(arg_num, Z_EXPECTED_LONG, arg ## arg_num); \
+		goto flf_clean; \
+	}
+#define Z_FLF_PARAM_LONG_OR_NULL(arg_num, is_null, dest) \
+	if (!zend_parse_arg_long_ex(arg ## arg_num, &dest, &is_null, /* null_check */ true, arg_num, /* frameless */ true)) { \
+		zend_wrong_parameter_type_error(arg_num, Z_EXPECTED_LONG_OR_NULL, arg ## arg_num); \
+		goto flf_clean; \
+	}
+#define Z_FLF_PARAM_STR(arg_num, dest, tmp) \
+	if (Z_TYPE_P(arg ## arg_num) == IS_STRING) { \
+		dest = Z_STR_P(arg ## arg_num); \
+	} else { \
+		ZVAL_COPY(&tmp, arg ## arg_num); \
+		arg ## arg_num = &tmp; \
+		if (!zend_parse_arg_str_ex(arg ## arg_num, &dest, /* null_check */ false, arg_num, /* frameless */ true)) { \
+			zend_wrong_parameter_type_error(arg_num, Z_EXPECTED_STRING, arg ## arg_num); \
+			goto flf_clean; \
+		} \
+	}
+#define Z_FLF_PARAM_FREE_STR(arg_num, tmp) \
+	if (UNEXPECTED(arg ## arg_num == &tmp)) { \
+		zval_ptr_dtor(arg ## arg_num); \
+	}
+
+BEGIN_EXTERN_C()
+
+typedef struct _zval_struct zval;
+typedef struct _zend_op zend_op;
+typedef union _zend_function zend_function;
+
+typedef void (*zend_frameless_function_0)(zval *return_value);
+typedef void (*zend_frameless_function_1)(zval *return_value, zval *op1);
+typedef void (*zend_frameless_function_2)(zval *return_value, zval *op1, zval *op2);
+typedef void (*zend_frameless_function_3)(zval *return_value, zval *op1, zval *op2, zval *op3);
+
+extern size_t zend_flf_count;
+extern size_t zend_flf_capacity;
+ZEND_API extern void **zend_flf_handlers;
+extern zend_function **zend_flf_functions;
+
+typedef struct {
+	void *handler;
+	uint32_t num_args;
+	/* Bit-list of arguments whose RC may be increased during the function call, for example when
+	 * returning the argument unmodified. In the JIT, arguments with inferred RC1 are normally freed
+	 * immediately using rc_dtor_func. With the corresponding flag set, they will use zval_ptr_dtor
+	 * instead. */
+	uint8_t arg_addref_info;
```

Comment:
I would rename `` arg_addref_info`` to ``flags`` and define corresponding constants (e.g. ZEND_FLF_ARG1_ADDREF).

Reply:
I think keeping flags separate (at least in the `@frameless-function` attribute) has some benefits. In particular, we can use sane default values. For `arg_addref_info`, the default value is `0b111` (i.e. we assume any param may increase the refcount). With `@frameless-function {"flags": "FOO"}` we can't know whether the user intended to express "this function doesn't addref to any args" or "i just didn't specify".

We could of course split the `@frameless-function` args but merge them in the flf info. If you prefer I can do so.

Reply:
I don't have any strong opinion, I just think this info is too low level for PHP extension APIs and the following code is too magic.

```php
/**
 * @compile-time-eval
 * @frameless-function {"arity": 1, "argAddrefInfo": "1"}
 * @frameless-function {"arity": 2, "argAddrefInfo": "10"}
 */
function trim(string $string, string $characters = " \n\r\t\v\0"): string {}
```

It would be great to improve this. I just gave an idea. For example the following code would look more clear.

```php
/**
 * @compile-time-eval
 * @frameless-function {"arity": 1, "mayCaptureArg1"}
 * @frameless-function {"arity": 2, "mayCaptureArg1"}
 */
function trim(string $string, string $characters = " \n\r\t\v\0"): string {}
```

Reply:
> I just think this info is too low level for PHP extension APIs and the following code is too magic.

I agree. Although, APIs don't *have* to provide this info, as this is more of a micro-optimization. I'll think about ways to improve this. I'll try to think about future flags and see if there's an approach that extends nicely.